src/io/mod.rs

```diff
@@ -1,5 +1,123 @@
 //! Input and output of images.
+
+use crate::{error, ImageError, ImageResult};
+
 mod reader;
 pub(crate) mod free_functions;
 
 pub use self::reader::Reader;
+
+/// Set of supported strict limits for a decoder
+#[derive(Clone, Debug, Eq, PartialEq, Hash)]
+#[allow(missing_copy_implementations)]
+pub struct LimitSupport {
+    _non_exhaustive: (),
+}
+
+impl Default for LimitSupport {
+    fn default() -> LimitSupport {
+        LimitSupport {
+            _non_exhaustive: (),
+        }
+    }
+}
+
+/// Resource limits for decoding.
+///
+/// Limits can be either *strict* or *non-strict*. Non-strict limits are best-effort
+/// limits where the library does not guarantee that limit wont be exceeded. Do note 
+/// that it is still considered a bug if a non-strict limit is exceeded, however as 
+/// some of the underlying decoders do not support not support such limits one cannot 
+/// rely on these limits being supported. For stric limits the library makes a stronger 
+/// guarantee that the limit will not be exceeded. Exceeding a strict limit is considered 
+/// a critical bug. If a decoder cannot guaranree that it will uphold a strict limit it 
+/// *must* fail with `image::error::LimitErrorKind::Unsupported`.
+///
+/// The limit check should only ever fail if a limit will be exceeded or an unsupported
+/// strict limit is used.
+#[derive(Clone, Debug, Eq, PartialEq, Hash)]
+#[allow(missing_copy_implementations)]
+pub struct Limits {
+    /// The maximum allowed image width. This limit is strict. The default is no limit.
+    pub max_image_width: Option<u32>,
+    /// The maximum allowed image height. This limit is strict. The default is no limit.
+    pub max_image_height: Option<u32>,
+    /// The maximum allowed amount of memory to be allocated by the 
+    /// decoder at any one time. This limit is non-strict and some 
+    /// decoders may ignore it. The default is 512MiB.
```

Comment:
This text is a bit ambiguous about whether "at one time" means that each individual allocation must be less than the limit, or whether the sum across all allocations that haven't been freed yet must be below the limit.